include/envoy/tracing/http_tracer.h

```diff
@@ -39,6 +39,40 @@ struct Decision {
   bool traced;
 };
 
+/**
+ * The context for the custom tag to obtain the tag value.
+ */
+struct CustomTagContext {
+  const Http::HeaderMap* request_headers;
+  const StreamInfo::StreamInfo& stream_info;
+};
+
+class Span;
+using SpanPtr = std::unique_ptr<Span>;
+
+/**
+ * Tracing custom tag, with tag name and how it would be applied to the span.
+ */
+class CustomTag {
+public:
+  virtual ~CustomTag() = default;
+
+  /**
+   * @return the tag name view.
+   */
+  virtual absl::string_view tag() const PURE;
+
+  /**
+   * The way how to apply the custom tag to the span,
+   * generally obtain the tag value from the context and attached it to the span.
+   * @param span the active span.
+   * @param ctx the custom tag context.
+   */
+  virtual void apply(Span& span, const CustomTagContext& ctx) const PURE;
+};
+
+using CustomTagPtr = std::shared_ptr<CustomTag>;
```

Comment:
nit: CustomTagSharedPtr, actually from the usage I think you want this be CustomTagConstSharedPtr no?

Reply:
Yeah, it should be a ConstSharedPtr